packages/analytics-core/src/plugins/destination.ts

```diff
@@ -297,21 +313,25 @@ export class Destination implements DestinationPlugin {
   }
 
   fulfillRequest(list: Context[], code: number, message: string) {
-    this.saveEvents();
     list.forEach((context) => context.callback(buildResult(context.event, code, message)));
   }
 
   /**
-   * Saves events to storage
-   * This is called on
-   * 1) new events are added to queue; or
-   * 2) response comes back for a request
+   * update the event storage
    */
-  saveEvents() {
+  async updateEvent(dropEventSet: Set<string>) {
     if (!this.config.storageProvider) {
       return;
     }
-    const events = Array.from(this.queue.map((context) => context.event));
-    void this.config.storageProvider.set(this.storageKey, events);
+    const savedEvent = await this.config.storageProvider.get(this.storageKey);
+    const reTriedEvents: Event[] = [];
+    if (savedEvent) {
+      savedEvent.filter((event) => {
+        if (!dropEventSet.has(event.event_type)) {
+          reTriedEvents.push(event);
+        }
+      });
+    }
+    void this.config.storageProvider.set(this.storageKey, reTriedEvents);
```

Comment:
Seems fine, just to confirm there wouldn't be any event go to storage after we read and before write?

Reply:
Added a lock to prevent the concurrency issue. Please check.